webapp/src/components/NavigateRegionsOverlay.tsx

```diff
@@ -0,0 +1,347 @@
+            return document.querySelector(".tutorialWrapper")?.getBoundingClientRect();
+        },
+        focus() {
+            findFirstFocusableDescendant(document.querySelector(".tutorialWrapper")!)?.focus();
+        }
+    },
+];
+
+export const NavigateRegionsOverlay = ({ parent }: NavigateRegionsOverlayProps) => {
+    const previouslyFocused = useRef<Element>(document.activeElement);
+    const getRects = (): Map<RegionId, DOMRect | undefined> => (
+        new Map(regions.map(region => [region.id, region.getBounds(parent)]))
+    );
+    const [regionRects, setRegionRects] = useState(getRects());
+
+    useEffect(() => {
+        if (parent.state.fullscreen) {
+            parent.setSimulatorFullScreen(false);
+        }
+
+        const listener = (e: KeyboardEvent) => {
+            const region = regions.find(region => region.shortcutKey === e.key);
+            if (region) {
+                e.preventDefault();
+                region.focus(parent);
+                parent.hideNavigateRegions();
+            }
+        }
+        document.addEventListener("keydown", listener)
+
+        const observer = new ResizeObserver(() => {
+            setRegionRects(getRects())
+        });
+        observer.observe(document.body);
+
+        return () => {
+            observer.disconnect()
+            document.removeEventListener("keydown", listener)
+        }
+    }, [])
+
+    const handleEscape = () => {
+        if (previouslyFocused.current) {
+            (previouslyFocused.current as HTMLElement).focus()
+        }
+        parent.hideNavigateRegions();
+    }
+
+    if (!regionRects.get("editor")) {
+        // Something is awry, bail out.
+        parent.hideNavigateRegions();
+        return null;
+    }
+
+    return ReactDOM.createPortal(
+        <FocusTrap dontRestoreFocus onEscape={handleEscape}>
+            <div className="navigate-regions-container">
+                {regions.map(region => {
+                    const rect = regionRects.get(region.id);
+                    return rect ? (<RegionButton
+                        key={region.id}
+                        title={region.ariaLabel}
+                        shortcutKey={region.shortcutKey}
+                        bounds={rect}
+                        onClick={() => {
+                            region.focus(parent);
+                            parent.hideNavigateRegions();
+                        }}
+                        ariaLabel={region.ariaLabel}
+                        className={region.getClassName?.(parent)}
+                    />) : null;
+                })}
+            </div>
+        </FocusTrap>,
+        document.getElementById("root") || document.body
+    );
+}
+
+interface RegionButtonProps extends ButtonProps {
+    shortcutKey: string;
+    bounds: RectBounds;
+}
+
+const RegionButton = ({ shortcutKey, bounds, ...props }: RegionButtonProps) => {
+    const buttonRef = useRef<HTMLButtonElement>()
+
+    useEffect(() => {
```

Comment:
is there a reason to use useEffect here? i think you could just set it in the style object in the JSX and let the usual props updating logic work

Reply:
Yep, thanks, changed as suggested in 52dc00c1334d5ab7070a2f45d0aa4de069437064

(Incidentally, I initially misinterpreted this and tried to remove the key handling part of the other useEffect, but that can't be done nicely as the FocusTrap div gets the focus but doesn't have an onKeyDown prop.)